### ISSUE TITLE
Cedar.EventStore.MsSql2008 doesn't store events in the correct order

### ISSUE THREAD

**Issue by bwaterschoot** [damianh/Cedar.EventStore#19]
What we're seeing is a mismatch between ordinal and streamversion for a stream.
Which leads to the highest ordinal number not having the highest streamversion.

![ordinal](https://cloud.githubusercontent.com/assets/3215301/13783574/da65b66e-eac3-11e5-9d58-549696e1d5de.png)

Full explantion and code: https://gist.github.com/bwaterschoot/77362620bbf058b58285

This happens because the events aren't stored in the correct order, link below:
https://social.msdn.microsoft.com/Forums/sqlserver/en-US/5cf485cc-a005-4f70-9cad-78b091d669cd/table-valued-parameters-order-of-records-to-stored-procedure?forum=transactsql

A solution would be in the ReadStreamForward to use StreamVersion instead of Ordinal as the orderby clause. Because you can't be sure SQL writes the event in the same order.

Probably would need to look at other Read sql statement to do something similar.

You can reproduce the issue using the code below, you might have to run it several times, because sometimes it does work ...

```c#    
    namespace Cedar.EventStore
    {
        using System;
        using System.Collections.Generic;
        using System.Threading.Tasks;
        using System.Linq;
        using Cedar.EventStore.Streams;
        using Shouldly;
        using Xunit;


        public partial class EventStoreAcceptanceTests
        {
        [Fact]
        public async Task Given_large_event_stream_can_be_read_back_in_pages()
        {
            using (var fixture = GetFixture("dbo"))
            {
                using (var eventStore = await fixture.GetEventStore())
                {
                    var eventsToWrite = CreateNewStreamEvents();

                    await eventStore.AppendToStream("stream-1", ExpectedVersion.NoStream, eventsToWrite);
                    

                    var readEvents = await new PagedEventStore(eventStore).GetAsync("stream-1");

                    readEvents.Count().ShouldBe(eventsToWrite.Length);
                }
            }
        }

        private NewStreamEvent[] CreateNewStreamEvents()
        {
            var eventsToWrite = new List<NewStreamEvent>();
            var largeStreamCount = 7500;
            for (int i = 0; i < largeStreamCount; i++)
            {
                var envelope = new NewStreamEvent(Guid.NewGuid(), $"event{i}", "{}", $"{i}");

                eventsToWrite.Add(envelope);
            }

            return eventsToWrite.ToArray();
        }
    }

    public class PagedEventStore
    {
        private readonly IEventStore _eventStore;

        public PagedEventStore(IEventStore eventStore)
        {
            _eventStore = eventStore;
        }

        public async Task<IEnumerable<StreamEvent>> GetAsync(string streamName)
        {
            var start = 0;
            const int BatchSize = 500;

            StreamEventsPage eventsPage;
            var events = new List<StreamEvent>();

            do
            {
                eventsPage = await _eventStore.ReadStreamForwards(streamName, start, BatchSize);

                if (eventsPage.Status == PageReadStatus.StreamDeleted)
                {
                    throw new Exception("Stream deleted");
                }

                if (eventsPage.Status == PageReadStatus.StreamNotFound)
                {
                    throw new Exception("Stream not found");
                }

                events.AddRange(
                    eventsPage.Events);

                start = eventsPage.NextStreamVersion;
            }
            while (!eventsPage.IsEndOfStream);

            return events;
        }

    }
}
```

**Comment by damianh** [comment]
> A solution would be in the ReadStreamForward to use StreamVersion instead of Ordinal as the orderby clause.

hmm, they should both be in the right order. It's needed when reading all streams.

**Comment by bwaterschoot** [comment]
Not sure whether you'll be able to pull that off:
http://blogs.msdn.com/b/sqltips/archive/2005/07/20/441053.aspx

**Comment by damianh** [comment]
Last resort is that I dump the TVP approach.

**Comment by yreynhout** [comment]
Isn't there a UC on the Ordinal column? Why not hand it out yourself (much like the StreamVersion)? I'm probably oversimplifying ...

**Comment by damianh** [comment]
@yreynhout it's an identity column https://github.com/damianh/Cedar.EventStore/blob/master/src/Cedar.EventStore.MsSql2008/SqlScripts/InitializeStore.sql#L13 . How would I do that from the client side? Only thing I can think of is some sort of counter that a client can 'take' x number of "ordinals"

**Comment by yreynhout** [comment]
Not from the client side, but inside the store/the script. Identity columns have been "replaced" by sequences in sql server anyway. Try it in Azure, you'll see the holes appear ...

**Comment by damianh** [comment]
> SQL Server 2012 introduced support for the sequence object as an alternative to the identity column property for generating surrogate keys. 

http://sqlmag.com/sql-server/sequence-and-identity-performance

Well aren't I out of date with sql server.

I want to support 2008(R2) though, does this look viable? https://blogs.msdn.microsoft.com/askjay/2012/10/08/sequence-objects-in-sql-2012-and-sql-2008/

**Comment by LeeCampbell** [comment]
I have provided a fairly simple fix in PR #20. It has been over 10 years since I wore a SQL hat, so would appreciate the criticism if it is not appropriate.
